### PR TITLE
mon/OSDMonitor: fix format error ceph osd stat --format json

### PR DESCRIPTION
**Pull request by zy751713126**
Fixes: https://tracker.ceph.com/issues/42501
Signed-off-by: Zheng Yin <zhengyin@cmss.chinamobile.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by zy751713126** [comment]
error output format:
[root@cephdev build]# ./bin/ceph osd stat
3 osds: 3 up (since 84s), 3 in (since 84s); epoch: e22

[root@cephdev build]#
[root@cephdev build]# ./bin/ceph osd stat --format json

223330

output format after fixing issue:

[root@cephdev build]# ./bin/ceph osd stat --format json

{"epoch":28,"num_osds":3,"num_up_osds":3,"num_in_osds":3,"num_remapped_pgs":0}

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -3962,11 +3962,13 @@ void OSDMap::print_summary(Formatter *f, ostream& out,
 			   const string& prefix, bool extra) const
 {
   if (f) {
+    f->open_object_section("osdmap");
```
Comment (tchaikov) [inline src/osd/OSDMap.cc]: this is not correct. normally, we don't print the enclosing "tags" in `print()` or `dump()` function. instead, we do this in the calling sites. see https://github.com/ceph/ceph/blob/626daf4b2043ede719a5d04c70468eb8bf282605/src/mon/Monitor.cc#L2945-L2947.

so i think a better fix is to update `OSDMonitor::preprocess_command()`.

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -5249,11 +5249,16 @@ bool OSDMonitor::preprocess_command(MonOpRequestRef op)
...
-    osdmap.print_summary(f.get(), ds, "", true);
-    if (f)
+    if (f) {
+      f->open_object_section("osdmap");
+      osdmap.print_summary(f.get(), ds, "", true);
+      f->close_section();
       f->flush(rdata);
-    else
+    }
+    else {
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: please

1. move `else {` into the previous line.
1. no need to pass `f.get()` to `osdmap.print_summary()` if `f` is empty. please just pass `nullptr` to it.

Comment (zy751713126) [reply src/mon/OSDMonitor.cc]: I have changed it, thanks for your good advice.

**Comment by tchaikov** [comment]
cc @ideepika

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-11-13_13:53:42-rados-wip-kefu-testing-2019-11-13-1728-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-11-15_11:39:33-rados-wip-kefu-testing-2019-11-13-1728-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-11-15_17:09:33-rados-wip-kefu-testing-2019-11-13-1728-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-11-16_10:02:59-rados-wip-kefu-testing-2019-11-16-1438-distro-basic-mira/

---

- the test.sh failures are fixed by #31689
- test_cls_rbd.sh already tracked by https://tracker.ceph.com/issues/42333, and it's fixed in latest master.
- test_diskprediction_local tracked by https://tracker.ceph.com/issues/42764

**Pull request merged**